src/ipc/ipc4/handler.c

```diff
@@ -0,0 +1,521 @@
+{
+	struct ipc4_module_large_config config;
+	struct comp_dev *dev;
+	int comp_id;
+	int ret;
+
+	memcpy_s(&config, sizeof(config), ipc4, sizeof(config));
+	tr_dbg(&ipc_tr, "ipc4_set_large_config_module_instance %x : %x with %x : %x",
+	       (uint32_t)config.header.r.module_id, (uint32_t)config.header.r.instance_id);
+
+	comp_id = IPC4_COMP_ID(config.header.r.module_id, config.header.r.instance_id);
+	dev = ipc4_get_comp_dev(comp_id);
+	if (!dev)
+		return IPC4_MOD_INVALID_ID;
+
+	ret = comp_cmd(dev, config.data.r.large_param_id, (void *)MAILBOX_HOSTBOX_BASE,
+		       config.data.dat);
+	if (ret < 0) {
+		tr_dbg(&ipc_tr, "failed to set large_config_module_instance %x : %x with %x : %x",
+		       (uint32_t)config.header.r.module_id, (uint32_t)config.header.r.instance_id);
+		ret = IPC4_INVALID_RESOURCE_ID;
+	}
+
+	return ret;
+}
+
+static int ipc4_delete_module_instance(union ipc4_message_header *ipc4)
+{
+	struct ipc4_module_delete_instance module;
+	struct ipc *ipc = ipc_get();
+	uint32_t comp_id;
+	int ret;
+
+	memcpy_s(&module, sizeof(module), ipc4, sizeof(module));
+	tr_dbg(&ipc_tr, "ipc4_delete_module_instance %x : %x", (uint32_t)module.header.r.module_id,
+	       (uint32_t)module.header.r.instance_id);
+
+	comp_id = IPC4_COMP_ID(module.header.r.module_id, module.header.r.instance_id);
+	ret = ipc_comp_free(ipc, comp_id);
+	if (ret < 0) {
+		tr_err(&ipc_tr, "failed to delete module instance %x : %x",
+		       (uint32_t)module.header.r.module_id,
+		       (uint32_t)module.header.r.instance_id);
+		ret = IPC4_INVALID_RESOURCE_ID;
+	}
+
+	return ret;
+}
+
+static int ipc4_process_module_message(union ipc4_message_header *ipc4)
+{
+	uint32_t type;
+	int ret;
+
+	type = ipc4->r.type;
+
+	switch (type) {
+	case SOF_IPC4_MOD_INIT_INSTANCE:
+		ret = ipc4_init_module_instance(ipc4);
+		break;
+	case SOF_IPC4_MOD_CONFIG_GET:
+	case SOF_IPC4_MOD_CONFIG_SET:
+	case SOF_IPC4_MOD_LARGE_CONFIG_GET:
+		ret = IPC4_UNAVAILABLE;
+		break;
+	case SOF_IPC4_MOD_LARGE_CONFIG_SET:
+		ret = ipc4_set_large_config_module_instance(ipc4);
+		break;
+	case SOF_IPC4_MOD_BIND:
+		ret = ipc4_bind_module_instance(ipc4);
+		break;
+	case SOF_IPC4_MOD_UNBIND:
+		ret = ipc4_unbind_module_instance(ipc4);
+		break;
+	case SOF_IPC4_MOD_DELETE_INSTANCE:
+		ret = ipc4_delete_module_instance(ipc4);
+		break;
+	case SOF_IPC4_MOD_SET_DX:
+	case SOF_IPC4_MOD_SET_D0IX:
+	case SOF_IPC4_MOD_ENTER_MODULE_RESTORE:
+	case SOF_IPC4_MOD_EXIT_MODULE_RESTORE:
+		ret = IPC4_UNAVAILABLE;
+		break;
+	default:
+		ret = IPC4_UNAVAILABLE;
+		break;
+	}
+
+	return ret;
+}
+
+ipc_cmd_hdr *mailbox_validate(void)
+{
+	ipc_cmd_hdr *hdr = ipc_get()->comp_data;
+	return hdr;
+}
+
+/*
+ * Most ABI 4 messages use compact format - keep logic simpler
+ * and handle all in IPC command.
+ */
+static uint32_t msg_in[2]; /* local copy of current message from host header */
+static uint32_t msg_out[2]; /* local copy of current message to host header */
```

Comment:
Lets do something better here. We should really have a structure for this, even if it's a structure that contains 2 uint32_t

Reply:
sure